internal/webhook/managedcluster_webhook.go

```diff
@@ -71,18 +72,35 @@ func (v *ManagedClusterValidator) ValidateCreate(ctx context.Context, obj runtim
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
-func (v *ManagedClusterValidator) ValidateUpdate(ctx context.Context, _ runtime.Object, newObj runtime.Object) (admission.Warnings, error) {
+func (v *ManagedClusterValidator) ValidateUpdate(ctx context.Context, oldObj runtime.Object, newObj runtime.Object) (admission.Warnings, error) {
+	oldManagedCluster, ok := oldObj.(*v1alpha1.ManagedCluster)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected ManagedCluster but got a %T", oldObj))
+	}
 	newManagedCluster, ok := newObj.(*v1alpha1.ManagedCluster)
 	if !ok {
 		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected ManagedCluster but got a %T", newObj))
 	}
-	template, err := v.getManagedClusterTemplate(ctx, newManagedCluster.Namespace, newManagedCluster.Spec.Template)
-	if err != nil {
-		return nil, fmt.Errorf("%s: %v", errInvalidManagedCluster, err)
-	}
-	err = v.isTemplateValid(ctx, template)
-	if err != nil {
-		return nil, fmt.Errorf("%s: %v", errInvalidManagedCluster, err)
+	oldTemplateName := oldManagedCluster.Spec.Template
+	newTemplateName := newManagedCluster.Spec.Template
+	if oldTemplateName != newTemplateName {
+		template, err := v.getManagedClusterTemplate(ctx, newManagedCluster.Namespace, newManagedCluster.Spec.Template)
+		if err != nil {
+			return nil, fmt.Errorf("%s: %v", errInvalidManagedCluster, err)
+		}
+		if !newManagedCluster.Spec.DryRun && templateutil.IsManagedByHMC(template) {
+			availableForUpgrade, err := templateutil.IsAvailableForUpgrade(ctx, v.Client, v1alpha1.ClusterTemplateKind, oldTemplateName, newTemplateName)
+			if err != nil {
+				return nil, fmt.Errorf("failed to check if the ClusterTemplate %s is available for the upgrade from %s: %w", newTemplateName, oldTemplateName, err)
+			}
+			if !availableForUpgrade {
+				return nil, fmt.Errorf("ClusterTemplate %s is not available for the upgrade from %s", newTemplateName, oldTemplateName)
+			}
+		}
+		err = v.isTemplateValid(ctx, template)
```

Comment:
validation might be executed before the check on the eligibility for the upgrade (technically it might save a couple of network calls)